pycalphad/mapping/zpf_equilibrium.py

```diff
@@ -0,0 +1,232 @@
+    orig_cs = [cs for cs in comp_sets]
+    try:
+        solver = Solver(remove_metastable=True, allow_changing_phases=False)
+        results = solver.solve(comp_sets, new_conditions)
+        if not results.converged:
+            return None
+    except Exception:
+        return None
+
+    # Add free variable back
+    if free_var is not None:
+        new_conditions[free_var] = free_var.compute_property(comp_sets, new_conditions, results.chemical_potentials)
+
+    new_point = Point(new_conditions, np.array(results.chemical_potentials), [cs for cs in comp_sets if cs.fixed], [cs for cs in comp_sets if not cs.fixed])
+    return new_point, orig_cs
+
+def find_global_min_point(point: Point, system_info: dict, pdens = 500, tol = 1e-5, num_candidates = 1):
+    """
+    For each possible phase:
+        1. Sample DOF and find CS that minimizes driving force
+        2. Create a DormantPhase with CS and compute driving force with potentials at equilibrium
+        3. If driving force is negative, then new phase is stable
+        4. Check that the new CS doesn"t match with a currently stable CS
+        4. Hope that this works on miscibility gaps
+
+    This should take care of the DLASCLS error since we compute the new phase separately so if
+    the composition clashes with a fixed phase, we check that afterwards before attempting to
+    run equilibrium on two CS with the same composition
+    """
+    dbf, comps, phases, models, phase_records = system_info["dbf"], system_info["comps"], system_info["phases"], system_info["models"], system_info["phase_records"]
+    # Get driving force and find index that maximized driving force
+    state_conds = {str(key): point.global_conditions[key] for key in STATEVARS}
+    points = calculate(dbf, comps, phases, model=models, phase_records=phase_records, output="GM", to_xarray=False, pdens=pdens, **state_conds)
+    gm = np.squeeze(points.GM)
+    x = np.squeeze(points.X)
+    y = np.squeeze(points.Y)
+    phase_ids = np.squeeze(points.Phase)
+    g_chempot = x * point.chemical_potentials
+    dGs = np.sum(g_chempot, axis=1) - gm
+
+    sorted_indices = np.argsort(dGs)[::-1]
+    cs = None
+    dG = 0
+    tested_phases = []
+    for i in range(np.amin([num_candidates, len(sorted_indices)])):
+        index = sorted_indices[i]
+        if phase_ids[index] not in tested_phases:
+            tested_phases.append(phase_ids[index])
+            test_cs = CompositionSet(phase_records[phase_ids[index]])
+            # TODO: on some models, I get an error of
+            # ValueError: buffer source array is read-only (line 59 in pycalphad.core.composition_set.CompositionSet.update)
+            # Databases where this occured in test suite: alcfe_b2.tdb running Al-C-Fe isopleth, alcocrni.tdb running Al-Co-Ni ternary
+            # and femns.tdb with ionic liquid model
```

Comment:
Resolve TODO if we can. I have a list of some binaries that trigger the exceptions, not sure if this one is included. 

This is a good thing to put in the test suite and fix to make sure that changes we make don't trigger known bugs in the future

Reply:
This seems to be due to the site fraction array from calculate being non-writable for some models. I'm not sure why this occurs, but the simple fix for mapping purposes is to create a new numpy array for the site fractions before updating the composition set since the new array will be writable.

Example of the site fractions array from femns.tdb being read-only.
```python
db = Database('femns.tdb')
comps = ['FE', 'MN', 'S', 'VA']
phases = list(db.phases.keys())
points = calculate(db, comps, phases, T=1000, P=101325, to_xarray=False)
y = np.squeeze(points.Y)
print('femns.tdb')
print('From np.squeeze: writeable =', y.flags.writeable)
y = np.array(y, dtype=np.float64)
print('From np.array writeable =', y.flags.writeable)

db = Database('crtiv_ghosh.tdb')
comps = ['CR', 'TI', 'V', 'VA']
phases = list(db.phases.keys())
points = calculate(db, comps, phases, T=1000, P=101325, to_xarray=False)
y = np.squeeze(points.Y)
print('\ncrtiv_ghosh.tdb')
print('From np.squeeze writeable =', y.flags.writeable)
y = np.array(y, dtype=np.float64)
print('From np.array writeable =', y.flags.writeable)
```

Output
```
femns.tdb
From np.squeeze: writeable = False
From np.array writeable = True

crtiv_ghosh.tdb
From np.squeeze writeable = True
From np.array writeable = True
```